Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/opensaml/CertificateUtil.java

```diff
@@ -76,4 +97,149 @@ private void closeStream(Closeable stream) {
+            cert = keystore.getCertificate(alias);
+            subjectKeyId = getCertKeyIdSubject(cert);
+            retObj.put(subjectKeyId, cert);
+        }
+        return retObj;
+    }
+
+    public static List<Certificate> getChain(Certificate cert, final KeyStore keystore) throws KeyStoreException {
+        return getChain((X509Certificate) cert, keystore);
+    }
+    public static List<Certificate> getChain(X509Certificate cert, final KeyStore keystore) throws KeyStoreException {
+        if (null == cert) {
+            return new ArrayList<>();
+        }
+        Map<String, Certificate> keyCache = getKeystoreMap(keystore);
+        List<Certificate> chain = new ArrayList<>();
+        chain.add(cert);
+
+        Certificate certLoop = cert;
+        String aki = getCertKeyIdAuthority(certLoop); // if-root: expected-null
+        while (null != aki) {
+            certLoop = keyCache.get(aki);
+            chain.add(certLoop);
+            aki = getCertKeyIdAuthority(certLoop);
+        }
+
+        return chain;
+    }
+
+    public static String getCertKeyIdAuthority(Certificate cert) {
+        return getCertKeyIdAuthority((X509Certificate) cert);
+    }
+    public static String getCertKeyIdAuthority(X509Certificate cert) {
+        if (null == cert) {
+            return null;
+        }
+        byte[] akiBytes = cert.getExtensionValue(AUTHORITY_KEY_ID);
+        String aki = null;
+        try {
+            if (akiBytes != null) {
+                DERDecoder extValA = new DERDecoder(akiBytes);
+                extValA.skip(6);
+                aki = Hex.encodeHexString(extValA.getBytes(20));
+            }
+        } catch (WSSecurityException e) {
+            LOG.error("Unable to convert AIK into human readable {}", e.getLocalizedMessage(), e);
+        }
+        return aki;
+    }
+
+    public static String getCertKeyIdSubject(Certificate cert) {
+        return getCertKeyIdSubject((X509Certificate) cert);
+    }
+    public static String getCertKeyIdSubject(X509Certificate cert) {
+        if (null == cert) {
+            return null;
+        }
+        byte[] subjectKeyID = cert.getExtensionValue(SUBJECT_KEY_ID);
+        String ski = null;
+        try {
+            if (subjectKeyID != null) {
+                // this logic extracts from CryptoBase class inside wss4j
+                DERDecoder extVal = new DERDecoder(subjectKeyID);
+                extVal.expect(DERDecoder.TYPE_OCTET_STRING); // ExtensionValue OCTET STRING
+                extVal.getLength();
+                extVal.expect(DERDecoder.TYPE_OCTET_STRING); // KeyIdentifier OCTET STRING
+                int keyIDLen = extVal.getLength();
+                byte[] hexValue = extVal.getBytes(keyIDLen);
+                ski = Hex.encodeHexString(hexValue);
+            }
+        } catch (WSSecurityException e) {
+            LOG.error("Unable to convert SKI into human readable {}", e.getLocalizedMessage(), e);
```

Comment:
What does this error mean?
"Unable to decode certificate subject" would be a a better/more meaningful error.